source/agora/common/Amount.d

```diff
@@ -0,0 +1,239 @@
+    }
+
+    /// Make a value without checking bounds
+    /// Used to initialize the bounds themselves, and to make invalid values
+    /// in unittests
+    private this (ulong units, bool dummy) { this.value = units; }
+    /// Ditto
+    version (unittest)
+    public static Amount invalid (ulong units) { return Amount(units, true); }
+
+    /***************************************************************************
+
+        Returns:
+            `true` if `value` is in the range `[0; Amount.MaxUnitSupply]`
+            and can be safely used to make an `Amount`
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public static bool isInRange (ulong value)
+    {
+        return  value <= MaxUnitSupply.value;
+    }
+
+    /// `isInRange` but as member function
+    pragma(inline, true)
+    public bool isValid () const
+    {
+        return isInRange(this.value);
+    }
+
+
+    /***************************************************************************
+
+        Add another value to `this` `Amount`
+
+        Params:
+            other = Another `Amount` value. If the value is invalid,
+                    `false` will be returned.
+        Returns:
+            `true` if the addition returned a number within bounds.
+            `false` if the number is out of the [0; MaxUnitSupply] bound.
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public bool add (Amount other)
+    {
+        bool overflow;
+        this.value = addu(this.value, other.value, overflow);
+        if (overflow || !this.isValid())
+        {
+            // If we overflow, make sure to poison the return value
+            this.value = ulong.max;
+            return false;
+        }
+        return true;
+    }
+
+    /***************************************************************************
+
+        Substract another value from `this` `Amount`
+
+        Params:
+            other = Another `Amount` value. If the value is invalid,
+                    `false` will be returned.
+        Returns:
+            `true` if the addition returned a number within bounds.
+            `false` if the number is out of the [0; MaxUnitSupply] bound.
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public bool sub (Amount other)
+    {
+        bool underflow;
+        this.value = subu(this.value, other.value, underflow);
+        if (underflow || !this.isValid())
+        {
+            // If we underflow, make sure to poison the return value
+            this.value = ulong.max;
+            return false;
+        }
+        return true;
+    }
+
+    /// Convenience version of `add` which asserts in case of overflow
+    /// Prefer using this only in `unittest`s
+    public ref Amount mustAdd (Amount other)
+    {
+        if (!this.add(other))
+            assert(0);
```

Comment:
Seems like it could be a DoS vulnerability to keep this public, if you're never supposed to use it. It crashes the app right..

Reply:
Yep it does. It is "practical" but should be flagged in any code review. I could also make it `version (unittest)` but I wanted to see what the diff would look like